node/src/chain/mod.rs

```diff
@@ -0,0 +1,70 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use crate::data::Topics;
+use crate::utils::PendingQueue;
+use crate::{data, database, Network};
+use crate::{LongLivedService, Message};
+use async_trait::async_trait;
+use tokio::sync::RwLock;
+
+use std::sync::Arc;
+
+const TOPICS: &[u8] = &[data::Topics::Block as u8];
+
+#[derive(Default)]
+pub struct ChainSrv {
+    inbound: PendingQueue<Message>,
+}
+
+#[async_trait]
+impl<N: Network, DB: database::DB> LongLivedService<N, DB> for ChainSrv {
+    async fn execute(
+        &mut self,
+        network: Arc<RwLock<N>>,
+        db: Arc<RwLock<DB>>,
+    ) -> anyhow::Result<usize> {
+        LongLivedService::<N, DB>::add_routes(
+            self,
+            TOPICS,
+            self.inbound.clone(),
+            &network,
+        )
+        .await?;
+
+        loop {
+            if let Ok(msg) = self.inbound.recv().await {
```

Comment:
@goshawk-3 Here's an example server that gracefully shuts down with a `tokio::select!` approach

https://github.com/tokio-rs/mini-redis/blob/master/src/server.rs#L174

Reply:
I couldn't find out any reasoning for opting for exit-chan approach. In Rust Tokio, aborting a task is much more faster as it uses any `await call` as a cancellation point. In other words, the implementation already uses a shutdown strategy which is based on handling Interrupt signal and aborting all running tasks more efficiently.

Reply:
In tokio, await is like a yield, when you reach an await you also push the future in the completion pool and then poll it to completion, if the future returns () and exits the branch of the server, then server gracefully shutsdown. The only reason thw shutdown works in the mini redis example is because of tokio::select!.
It returns the shutdown branch if we recv signal, else it keeps polling the server for requests